# Анализ вакансий с сайта LinkedIn. Подготовка к визуализации

## Описание данных

### Описание проекта

Для анализа представлены данные с информацией о рынке европейских вакансий для дата аналитика / BI аналитика.  
Данные представлены виде файла .сsv c вакансиями, спарсенными с LinkedIn 07/09/2022 за неделю.  

### Цели и задачи

В рамках проекта необходимо провести анализ вакансий, для чего выполнить задачи:  
1. распарсить csv файл и выделить характерные признаки для каждой вакансии в понятный датафрейм (название вакансии, местоположение, данные о работодателе, тип занятости и т.п.),
2. провести предварительную обработку данных,
3. провести анализ данных с визуализацией и построение дашборда.

В рамках настоящего проекта будут выполнены задачи №№ 1-2. Задача визуализации будет выполнена в Tableau.

## Импорт библиотек и загрузка данных

In [1]:
import pandas as pd
from bs4 import BeautifulSoup as bs
import numpy as np
from datetime import datetime, timedelta
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
pd.set_option('display.max_colwidth', 70)

In [4]:
df = pd.read_csv('masterskaya_yandex_2022_09_07.csv', index_col=0)
df['html'][0]

'\n  <div>\n    <div class="\n        jobs-details__main-content jobs-details__main-content--single-pane full-width\n        ">\n        \n<!---->\n      \n<!---->\n      <div>\n        <div class="jobs-unified-top-card t-14">\n  <div class="relative">\n\n    <div class="jobs-unified-top-card__content--two-pane">\n<!---->\n        <a href="/jobs/view/3258155313/?alternateChannel=search&amp;refId=xQWMq5sHxPa4BOVKDp0TKg%3D%3D&amp;trackingId=grGqx%2FHW0EZ%2BorBvEr%2Ft%2Bg%3D%3D&amp;trk=d_flagship3_search_srp_jobs" id="ember464" class="ember-view">\n          <h2 class="t-24 t-bold jobs-unified-top-card__job-title">Stage - Assistant Ingénieur Qualité - Beyrand (H/F)</h2>\n        </a>\n\n      <div class="jobs-unified-top-card__primary-description">\n          <span class="jobs-unified-top-card__subtitle-primary-grouping t-black">\n            <span class="jobs-unified-top-card__company-name">\n                  <a href="/company/hermes-group/life/" id="ember465" class="ember-view t-black 

## Распарсивание данных

### Навазние вакансии

Выведем из html названия вакансий

In [5]:
df['title'] = df['html'].apply(lambda x: bs(x).find('h2').text.strip())

In [6]:
df.head()

html  \
0  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
3  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
4  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                 title  
0  Stage - Assistant Ingénieur Qualité - Beyrand (H/F)  
1  développeur matlab/simulink, secteur automobile f/h  
2                                  Online Data Analyst  
3                        Online Data Analyst - Belgium  
4                                         Data Analyst

### Местоположение вакансии

Далее распарсим местоположение вакансии

In [7]:
def get_place(cell):
    try:
        return bs(cell).find('span', class_='jobs-unified-top-card__bullet').text.strip()
    except:
        np.nan

df['place'] = df['html'].apply(get_place)

In [8]:
df.head()

html  \
0  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
3  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
4  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                 title  \
0  Stage - Assistant Ingénieur Qualité - Beyrand (H/F)   
1  développeur matlab/simulink, secteur automobile f/h   
2                                  Online Data Analyst   
3                        Online Data Analyst - Belgium   
4                                         Data Analyst   

                                    place  
0     Limoges, Nouvelle-Aquitaine, France  
1             Toulouse, Occitanie, France  
2   Skara, Vastra Gotaland County, Sweden  
3  West Flanders, Flemish Region, Belgium  
4     Mecklenburg-West Pomerania, Germany

Из полученного местоположения выделим в отдельные столбцы название города и страну

In [9]:
# город

def get_city(cell):
    if len(cell.split(',')) > 1:
        return cell.split(',')[0].strip()
    elif "Metropolitan" in cell or "Greater" in cell:
        return cell.replace('Greater', '').replace('Metropolitan', '').replace('Area', '').replace('Region','').strip()
    else:
        return np.nan

df['city'] = df['place'].apply(get_city)

In [10]:
# страна

def get_country(cell):
    if len(cell.split(',')) > 1:
        return cell.split(',')[-1].strip()
    elif "Metropolitan" in cell or "Greater" in cell or "Region" in cell:
        return np.nan
    else:
        return cell
    
df['country'] = df['place'].apply(get_country)

In [11]:
df.head()

html  \
0  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
3  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
4  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                 title  \
0  Stage - Assistant Ingénieur Qualité - Beyrand (H/F)   
1  développeur matlab/simulink, secteur automobile f/h   
2                                  Online Data Analyst   
3                        Online Data Analyst - Belgium   
4                                         Data Analyst   

                                    place                        city  country  
0     Limoges, Nouvelle-Aquitaine, France                     Limoges   France  
1             Toulouse, Occitanie, France                    Toulouse   France  
2   Skara, Vastra Gotaland County, Sweden                       Skara   Sweden  
3  West Flanders, Flemish Region, Belgium               West Flanders  Belgium  
4     Mecklenburg-West Pomerania, Germany  Mecklenburg-West Pomerania  Germany

### Тип занятости

Далее выделим тип занятости

In [12]:
def work_type(cell):
    try:
        return bs(cell).find('span', class_='jobs-unified-top-card__workplace-type').text.strip()
    except:
        np.nan

df['work_type'] = df['html'].apply(work_type)


### Данные о компании-работодателе

Найдем и сохраним в отдельные столбцы данные о компании - её название, размер (количество сотрудников) и сферу деятельности

In [13]:
# название компании

def company_name(cell):
    try:
        return bs(cell).find('span', class_='jobs-unified-top-card__company-name').text.strip()
    except:
        np.nan

df['company_name'] = df['html'].apply(company_name)

In [14]:
# размер компании

def company_size(cell):
    try:
        return bs(cell).find('div', class_='mt5 mb2').find_all('li', class_='jobs-unified-top-card__job-insight'
                                               )[1].text.split(' · '
                                                              )[0].replace('\n', '').replace(' employees', '')
    except:
        np.nan

df['company_size'] = df['html'].apply(company_size)

In [15]:
# сфера деятельности компании

def company_area(cell):
    try:
        return bs(cell).find('div', class_='mt5 mb2').find_all('li', class_='jobs-unified-top-card__job-insight'
                                               )[1].text.split(' · '
                                                              )[1].replace('\n', '')
    except:
        np.nan

df['company_area'] = df['html'].apply(company_area)

In [16]:
df.head()

html  \
0  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
3  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
4  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                 title  \
0  Stage - Assistant Ingénieur Qualité - Beyrand (H/F)   
1  développeur matlab/simulink, secteur automobile f/h   
2                                  Online Data Analyst   
3                        Online Data Analyst - Belgium   
4                                         Data Analyst   

                                    place                        city  \
0     Limoges, Nouvelle-Aquitaine, France                     Limoges   
1             Toulouse, Occitanie, France                    Toulouse   
2   Skara, Vastra Gotaland County, Sweden                       Skara   
3  West Flanders, Flemish Region, Belgium               West Flanders   
4     Mecklenburg-West Pomerania, Germany  Mecklenburg-West Pomerania   

   country work_type                           company_name  company_size  \
0   France   On-site                                 Hermès       10,001+   
1   France   On-site                                   AUSY  5,001-10,000   
2   Sweden    Remote  TELUS International AI Data Solutions       10,001+   
3  Belgium    Remote                    TELUS International       10,001+   
4  Germany    Remote  TELUS International AI Data Solutions       10,001+   

                      company_area  
0  Retail Luxury Goods and Jewelry  
1    IT Services and IT Consulting  
2    IT Services and IT Consulting  
3    IT Services and IT Consulting  
4    IT Services and IT Consulting

### Hard-skills

Далее выведем в отдельный столбец описание вакансии, чтобы из него выделить hard-skills

In [17]:
# распарсиваем описание вакансии

df['description'] = df['html'].apply(lambda x: bs(x).find('div', {'id':'job-details'}).text.strip())

In [18]:
# создаем список hard-skills

skills = (['datahub', 'api', 'github', 'google analytics', 'adobe analytics', 'ibm coremetrics', 'omniture',
            'gitlab', 'erwin', 'hadoop', 'spark', 'hive', 'databricks', 'aws', 'gcp', 'azure','excel',
            'redshift', 'bigquery', 'snowflake',  'hana', 'grafana', 'kantar', 'spss', 
           'asana', 'basecamp', 'jira', 'dbeaver','trello', 'miro', 'salesforce', 
           'rapidminer', 'thoughtspot',  'power point',  'docker', 'jenkins','integrate.io', 'talend', 'apache nifi',
          'aws glue','pentaho','google data flow', 'azure data factory','xplenty','skyvia','iri voracity','xtract.io',
          'dataddo', 'ssis', 'hevo data','informatica','oracle data integrator','k2view','cdata sync','querysurge', 
             'rivery', 'dbconvert', 'alooma', 'stitch', 'fivetran', 'matillion','streamsets','blendo',
             'iri voracity','logstash', 'etleap', 'singer', 'apache camel','actian', 'airflow', 'luidgi', 'datastage',
           'python', 'vba', 'scala', ' r ', 'java script', 'julia', 'sql', 'matlab', 'java', 'html', 'c++', 'sas',
           'data studio', 'tableau', 'looker', 'powerbi', 'cognos', 'microstrategy', 'spotfire',
             'sap business objects','microsoft sql server', 'oracle business intelligence', 'yellowfin',
             'webfocus','sas visual analytics', 'targit', 'izenda',  'sisense', 'statsbot', 'panorama', 'inetsoft',
             'birst', 'domo', 'metabase', 'redash', 'power bi', 'alteryx', 'dataiku', 'qlik sense', 'qlikview', 
         'hypotheses', 'ab testing', 'a/b testing', 'ab tests', 'a/b tests', 'statistics', 'statistical analysis'])

In [19]:
# с помощью функции проверяем наличие hard-skills из списка в описании вакансии 
# и сохраняем найденные hard-skills в отдельный столбец

def hard_skills(cell):
    list_skills = []
    for skill in skills:
        if skill in cell.lower().replace('powerbi', 'power bi'):
            list_skills.append(skill)
    return list_skills

df['hard_skills'] = df['description'].apply(hard_skills)

In [20]:
df.head(3)

html  \
0  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                 title  \
0  Stage - Assistant Ingénieur Qualité - Beyrand (H/F)   
1  développeur matlab/simulink, secteur automobile f/h   
2                                  Online Data Analyst   

                                   place      city country work_type  \
0    Limoges, Nouvelle-Aquitaine, France   Limoges  France   On-site   
1            Toulouse, Occitanie, France  Toulouse  France   On-site   
2  Skara, Vastra Gotaland County, Sweden     Skara  Sweden    Remote   

                            company_name  company_size  \
0                                 Hermès       10,001+   
1                                   AUSY  5,001-10,000   
2  TELUS International AI Data Solutions       10,001+   

                      company_area  \
0  Retail Luxury Goods and Jewelry   
1    IT Services and IT Consulting   
2    IT Services and IT Consulting   

                                                             description  \
0  LA SOCIETE : \nCréée en 1926, la société Beyrand, filiale du Group...   
1  Dans le cadre de la croissance de nos activités automobile, nous r...   
2  TELUS International AI-Data Solutions partners with a diverse and ...   

    hard_skills  
0  [api, excel]  
1      [matlab]  
2            []

### Иные сведения о вакансии

Также выделим из html данные о дате публикации вакансии и количества откликнувшихся кандидатов

In [21]:
# дата публикации

def publish_date(cell):
    try:
        return bs(cell).find('span', class_='jobs-unified-top-card__posted-date').text.strip()
    except:
        np.nan

df['publish_date'] = df['html'].apply(publish_date)

In [22]:
# количество кандидатов

def get_candidats(cell):
    try:
        return bs(cell).find('span', class_='jobs-unified-top-card__applicant-count'
                            ).text.strip().replace('applicants', '').replace('applicant', '')
    except:
        pd.NA
        
df['candidats'] = df['html'].apply(get_candidats)

## Предобработка данных

### Фильтр по вакансиям

Отфильтруем датасет по релевантным вакансиям

In [23]:
# создадим список релевантных вакансий и проверим совпадение вакансий в столбце title с релевантными вакансиями из списка

relevant_vac = ['Data Analyst', 'Data analyst', 'data analyst', ' Data Analyst' 
                'Product Analyst', 'Product analyst', 'product analyst', 
                'Business Intelligence', 'Business intelligence', 'business intelligence']

vac = []
for i in df['title']:
    for k in relevant_vac:
        if k in i:
            vac.append(i)

In [24]:
# сделаем срез по релевантным вакансиям и сохраним в отдельный датасет

df_rel = df.loc[df['title'].isin(vac)].reset_index(drop=True)

In [25]:
df_rel.shape

(262, 13)

Итого после фильтрации релевантных вакансий у нас осталось 262 строки из 772.

### Заполнение пропусков

Проверим полученный датасет на наличие пропусков

In [26]:
df_rel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   html          262 non-null    object
 1   title         262 non-null    object
 2   place         262 non-null    object
 3   city          245 non-null    object
 4   country       250 non-null    object
 5   work_type     213 non-null    object
 6   company_name  262 non-null    object
 7   company_size  261 non-null    object
 8   company_area  243 non-null    object
 9   description   262 non-null    object
 10  hard_skills   262 non-null    object
 11  publish_date  262 non-null    object
 12  candidats     234 non-null    object
dtypes: object(13)
memory usage: 26.7+ KB


In [27]:
df_rel[df_rel['city'].isna() == True]

html  \
9    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
12   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
31   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
70   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
82   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
83   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
93   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
110  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
126  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
158  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
183  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
204  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
205  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
211  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
214  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
217  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
234  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                    title           place city  \
9                  Data Analyst (Tableau)       Gibraltar  NaN   
12                           Data Analyst          Poland  NaN   
31          Online Data Analyst - Belgium         Belgium  NaN   
70                           Data Analyst          Poland  NaN   
82                    Online Data Analyst         Finland  NaN   
83                           Data Analyst         Germany  NaN   
93          Work From Home | Data Analyst          Poland  NaN   
110                          Data Analyst         Germany  NaN   
126              Data Analyst - Part time          Poland  NaN   
158  Data Analyst (Tijdelijk - 3 Maanden)     Netherlands  NaN   
183          Online Data Analyst - Sweden          Sweden  NaN   
204                          Data Analyst  United Kingdom  NaN   
205          Online Data Analyst - Sweden          Sweden  NaN   
211        Online Data Analyst - Slovenia        Slovenia  NaN   
214         Online Data Analyst - Croatia         Croatia  NaN   
217            Cybersecurity Data Analyst     Switzerland  NaN   
234  Data Analyst for Portuguese Speakers        Portugal  NaN   

            country work_type                           company_name  \
9         Gibraltar      None                          Guardian Jobs   
12           Poland    Remote  TELUS International AI Data Solutions   
31          Belgium    Remote                    TELUS International   
70           Poland    Remote  TELUS International AI Data Solutions   
82          Finland    Remote                    TELUS International   
83          Germany    Remote  TELUS International AI Data Solutions   
93           Poland    Remote  TELUS International AI Data Solutions   
110         Germany    Remote  TELUS International AI Data Solutions   
126          Poland    Remote  TELUS International AI Data Solutions   
158     Netherlands    Remote                 Independent Recruiters   
183          Sweden    Remote                    TELUS International   
204  United Kingdom    Remote                               Endeavor   
205          Sweden    Remote                    TELUS International   
211        Slovenia    Remote                    TELUS International   
214         Croatia    Remote                    TELUS International   
217     Switzerland    Remote                        Nozomi Networks   
234        Portugal    Remote                                  Appen   

     company_size                   company_area  \
9          51-200        Staffing and Recruiting   
12        10,001+  IT Services and IT Consulting   
31        10,001+  IT S

Пропуски в столбце city заполнить нельзя из-за отсутствия информации в исходных данных, поэтому оставляем NaN.

In [28]:
df_rel[df_rel['country'].isna() == True]

html  \
5    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
13   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
16   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
17   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
19   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
20   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
40   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
60   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
96   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
99   \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
170  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
179  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                            title  \
5                                                    Data Analyst   
13                                                   Data Analyst   
16                            Online Data Analyst | Flexible Work   
17                                                   Data Analyst   
19                                                   Data Analyst   
20                       Online Data Analyst | REMOTE OPPORTUNITY   
40                                           Remote| Data Analyst   
60                                  Work From Home | Data Analyst   
96   Data Engineer for Business Intelligence/ Data Analytics team   
99                                            Junior Data Analyst   
170                                                  Data Analyst   
179                                                  Data Analyst   

                                    place        city country work_type  \
5              Przemyśl Metropolitan Area    Przemyśl     NaN    Remote   
13            Grudziadz Metropolitan Area   Grudziadz     NaN    Remote   
16   Greater Norrköping Metropolitan Area  Norrköping     NaN    Remote   
17                Radom Metropolitan Area       Radom     NaN    Remote   
19               Zamosc Metropolitan Area      Zamosc     NaN    Remote   
20                 Greater Edinburgh Area   Edinburgh     NaN    Remote   
40                 Lodz Metropolitan Area        Lodz     NaN    Remote   
60                 Lodz Metropolitan Area        Lodz     NaN    Remote   
96            Bucharest Metropolitan Area   Bucharest     NaN    Hybrid   
99        Greater Milan Metropolitan Area       Milan     NaN    Hybrid   
170                   Greater Zurich Area      Zurich     NaN    Remote   
179   Greater Barcelona Metropolitan Area   Barcelona     NaN    Remote   

                              company_name company_size  \
5    TELUS International AI Data Solutions      10,001+   
13   TELUS International AI Data Solutions      10,001+   
16   TELUS International AI Data Solutions      10,001+   
17   TELUS International AI Data Solutions      10,001+   
19   TELUS International AI Data Solutions      10,001+   
20   TELUS International AI Data Solutions      10,001+   
40   TELUS International AI Data Solutions      10,001+   
60   TELUS International AI Data Solutions      10,001+   
96      LSEG (London Stock Exchange Group)      10,001+   
99                               Pardgroup    501-1,000   
170                                  Appen    501-1,000   
179                                  Appen    501-1,000   

                      company_area  \
5    IT Services and IT Consulting   
13   IT Services and IT Consulting   
16   IT Services and IT Consulting   
17   IT Services and IT Consulting   
19   IT Services and IT Consulting   
20   IT Services and IT Consulting   
40   IT Services and IT Consulting   
60   IT Services and IT Consulting   
96              Financial Services   
99            Advertising Service

Пропуски в странах заполним вручную на основе информации о городах:

In [29]:
df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Przemyśl', 'Grudziadz', 'Radom', 'Zamosc', 'Lodz'])), 'country'] = 'Poland'
        
df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Norrköping'])), 'country'] = 'Sweden'
        
df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Edinburgh'])), 'country'] = 'United Kingdom'
        
df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Bucharest'])), 'country'] = 'Romania'

df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Milan'])), 'country'] = 'Italy'
        
df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Zurich'])), 'country'] = 'Switzerland'
        
df_rel.loc[(df_rel['country'].isna() == True) & (df_rel['city'].isin(
        ['Barcelona'])), 'country'] = 'Spain'

In [30]:
df_rel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   html          262 non-null    object
 1   title         262 non-null    object
 2   place         262 non-null    object
 3   city          245 non-null    object
 4   country       262 non-null    object
 5   work_type     213 non-null    object
 6   company_name  262 non-null    object
 7   company_size  261 non-null    object
 8   company_area  243 non-null    object
 9   description   262 non-null    object
 10  hard_skills   262 non-null    object
 11  publish_date  262 non-null    object
 12  candidats     234 non-null    object
dtypes: object(13)
memory usage: 26.7+ KB


Пропусков в данных о типе занятости немного, поэтому оставим соответствующие поля пустыми.

В столбце company_size некоторые данные заполнены некорректно:

In [31]:
df_rel.company_size.value_counts()

10,001+                                                                                             108
1,001-5,000                                                                                          34
51-200                                                                                               25
201-500                                                                                              22
501-1,000                                                                                            19
5,001-10,000                                                                                         16
1-10                                                                                                 14
11-50                                                                                                12
See recent hiring trends for Software * IT. Unlock more Premium insights for free                     2
See recent hiring trends for www.TeamQuest.pl. Unlock more Premi

В связи с отсутствием иной информации заменим некорректно заполненные 11 значений в столбце company_size на 'no information' - тем самым добавятся пропуски к датафрейму, но данные будут корректными:

In [32]:
df_rel.loc[df_rel['company_size'].isin(['10,001+', '1,001-5,000', '51-200', '201-500', '501-1,000', 
                                        '5,001-10,000', '1-10', '11-50']) == False, 'company_size'] = 'no information'

In [33]:
df_rel.company_size.value_counts()

10,001+           108
1,001-5,000        34
51-200             25
201-500            22
501-1,000          19
5,001-10,000       16
1-10               14
11-50              12
no information     12
Name: company_size, dtype: int64

Пропуски в столбце company_area, company_size и candidats иным способом заполнить не представляется возможным, поэтому оставим их пустыми, тем более что количество пропусков не очень значительное.

### Расчёт даты публикации вакансии

Даты публикации нам даны не в виде дат, а в виде промежутков типа '6 дней назад'. Для последующего анализа необходимо расчитать конретные даты публикаций. Считать будем от даты парсинга - '2022-09-07 17:00:00'.

Для начала обозначим переменной дату парсинга, от которой будем отсчитывать даты публикации вакансий

In [34]:
parsing_date = pd.to_datetime('2022-09-07 17:00:00')

In [35]:
df_rel.publish_date.value_counts()

1 day ago       60
6 days ago      59
5 days ago      46
2 days ago      29
8 hours ago     13
4 days ago      11
23 hours ago     7
2 hours ago      7
3 days ago       6
7 hours ago      5
9 hours ago      4
21 hours ago     4
3 hours ago      2
14 hours ago     1
19 hours ago     1
4 hours ago      1
6 hours ago      1
18 hours ago     1
16 hours ago     1
11 hours ago     1
1 week ago       1
10 hours ago     1
Name: publish_date, dtype: int64

Как видим из имеющихся промежутков у нас есть дни, часы и недели. С помощью функции проверим вхождение соответствующих слов (день, час, неделя) дате публикации каждой из вакансий и отнимем от даты парсинга соответствующее значение, после чего сохраним полученную дату в отдельный столбец date.

In [36]:
def date(cell):
    if 'hours ago' in cell:
        cell = (parsing_date - timedelta(hours=int(cell.strip()[0]))).strftime('%Y-%m-%d')
        return cell
    
    elif 'days ago' in cell or 'day ago' in cell:
        cell = (parsing_date - timedelta(days=int(cell.strip()[0]))).strftime('%Y-%m-%d')
        return cell
    
    elif 'week ago' in cell:
        cell = (parsing_date - timedelta(weeks=int(cell.strip()[0]))).strftime('%Y-%m-%d')
        return cell
    
    else:
        return np.nan

In [37]:
df_rel['date'] = df_rel['publish_date'].apply(date)

In [38]:
df_rel.head()

html  \
0  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
3  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
4  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                    title  \
0                     Online Data Analyst   
1           Online Data Analyst - Belgium   
2                            Data Analyst   
3                            Data Analyst   
4  Alternant/ Alternante data analyst M/F   

                                         place                        city  \
0        Skara, Vastra Gotaland County, Sweden                       Skara   
1       West Flanders, Flemish Region, Belgium               West Flanders   
2          Mecklenburg-West Pomerania, Germany  Mecklenburg-West Pomerania   
3                             Hamburg, Germany                     Hamburg   
4  Rousset, Provence-Alpes-Côte d'Azur, France                     Rousset   

   country work_type                           company_name company_size  \
0   Sweden    Remote  TELUS International AI Data Solutions      10,001+   
1  Belgium    Remote                    TELUS International      10,001+   
2  Germany    Remote  TELUS International AI Data Solutions      10,001+   
3  Germany    Remote  TELUS International AI Data Solutions      10,001+   
4   France   On-site                     STMicroelectronics      10,001+   

                    company_area  \
0  IT Services and IT Consulting   
1  IT Services and IT Consulting   
2  IT Services and IT Consulting   
3  IT Services and IT Consulting   
4    Semiconductor Manufacturing   

                                                             description  \
0  TELUS International AI-Data Solutions partners with a diverse and ...   
1  TELUS International AI-Data Solutions partners with a diverse and ...   
2  TELUS International AI-Data Solutions partners with a diverse and ...   
3  TELUS International AI-Data Solutions partners with a diverse and ...   
4  POURQUOI NOUS REJOINDRE\n\nAu sein de ST, nous sommes 48000 hommes...   

                                            hard_skills publish_date  \
0                                                    []   6 days ago   
1                                                    []   6 days ago   
2                                                    []  8 hours ago   
3                                                    []  8 hours ago   
4  [excel, python, sql, java, html, spotfire, power bi]   5 days ago   

  candidats        date  
0       12   2022-09-01  
1       11   2022-09-01  
2        2   2022-09-07  
3        4   2022-09-07  
4       25   2022-09-02

### Проверка на дубликаты

In [39]:
print('Явных дубликатов:', df_rel.duplicated(subset=['html']).sum())

Явных дубликатов: 0


Теперь проверим дубликаты в сочетании 'название вакансии - компания - город - описание':

In [40]:
df_rel[df_rel.duplicated(subset=['title', 'company_name', 'city', 'description'])]

html  \
61  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
66  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
68  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                         title                                        place  \
61  Data Analyst  - Boursorama  Boulogne-Billancourt, Île-de-France, France   
66  Data Analyst  - Boursorama  Boulogne-Billancourt, Île-de-France, France   
68  Data Analyst  - Boursorama  Boulogne-Billancourt, Île-de-France, France   

                    city country work_type company_name company_size  \
61  Boulogne-Billancourt  France    Remote   Boursorama    501-1,000   
66  Boulogne-Billancourt  France    Remote   Boursorama    501-1,000   
68  Boulogne-Billancourt  France    Remote   Boursorama    501-1,000   

          company_area  \
61  Financial Services   
66  Financial Services   
68  Financial Services   

                                                              description  \
61  Avec son double positionnement unique de banque-média, Boursorama ...   
66  Avec son double positionnement unique de banque-média, Boursorama ...   
68  Avec son double positionnement unique de banque-média, Boursorama ...   

                      hard_skills publish_date candidats        date  
61  [excel, python, sas, tableau]   6 days ago       62   2022-09-01  
66  [excel, python, sas, tableau]   6 days ago       63   2022-09-01  
68  [excel, python, sas, tableau]   6 days ago       38   2022-09-01

Удалим дубликаты

In [41]:
df_rel = df_rel.drop_duplicates(subset=['title', 'company_name', 'city', 'description'], keep='last').reset_index(drop=True)
df_rel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   html          259 non-null    object
 1   title         259 non-null    object
 2   place         259 non-null    object
 3   city          242 non-null    object
 4   country       259 non-null    object
 5   work_type     210 non-null    object
 6   company_name  259 non-null    object
 7   company_size  259 non-null    object
 8   company_area  240 non-null    object
 9   description   259 non-null    object
 10  hard_skills   259 non-null    object
 11  publish_date  259 non-null    object
 12  candidats     231 non-null    object
 13  date          259 non-null    object
dtypes: object(14)
memory usage: 28.5+ KB


### Замена типов данных

In [42]:
df_rel.dtypes

html            object
title           object
place           object
city            object
country         object
work_type       object
company_name    object
company_size    object
company_area    object
description     object
hard_skills     object
publish_date    object
candidats       object
date            object
dtype: object

Подавляющее число данных у нас относятся к типу object - из них только данные в столбце candidats можем привести к числовому типу, а в столбце date - к datetime:

In [43]:
df_rel['candidats'] = df_rel['candidats'].astype('float').astype('Int64')
df_rel['date'] = pd.to_datetime(df_rel['date'])

Столбец с hard-skills разбиваем по каждому отдельному hard-skill для корректного построения дашборда.

In [44]:
df_rel = df_rel.reset_index()

In [45]:
df_rel

index  \
0        0   
1        1   
2        2   
3        3   
4        4   
..     ...   
254    254   
255    255   
256    256   
257    257   
258    258   

                                                                      html  \
0    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
1    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
2    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
3    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
4    \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
..                                                                     ...   
254  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
255  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
256  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
257  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   
258  \n  <div>\n    <div class="\n        jobs-details__main-content jo...   

                                                                     title  \
0                                                      Online Data Analyst   
1                                            Online Data Analyst - Belgium   
2                                                             Data Analyst   
3                                                             Data Analyst   
4                                   Alternant/ Alternante data analyst M/F   
..                                                                     ...   
254                                                           Data Analyst   
255                                         BI Data Analyst Intern (f/m/x)   
256                                                  Language Data Analyst   
257  Assistenz (w/m/d) des CEO West - Schwerpunkt Data Analyst / Contro...   
258                               Corporate Business Intelligence Engineer   

                                           place                        city  \
0          Skara, Vastra Gotaland County, Sweden                       Skara   
1         West Flanders, Flemish Region, Belgium               West Flanders   
2            Mecklenburg-West Pomerania, Germany  Mecklenburg-West Pomerania   
3                               Hamburg, Germany                     Hamburg   
4    Rousset, Provence-Alpes-Côte d'Azur, France                     Rousset   
..                                           ...                         ...   
254               Dublin, County Dublin, Ireland                      Dublin   
255                      Berlin, Berlin, Germany                      Berlin   
256                   Cork, County Cork, Ireland                        Cork   
257  Düsseldorf, North Rhine-Westphalia, Germany                  Düsseldorf   
258                 Paris, Île-de-France, France                       Paris   

     country work_type                           company_name company_size  \
0     Sweden    Remote  TELUS International AI Data Solutions      10,001+   
1    Belgium    Remote                    TELUS International      10,001+   
2    Germany    Remote  TELUS International AI Data Solutions      10,001+   
3    Germany    Remote  TELUS International AI Data Solutions      10,001+   
4     France   On-site                     STMicroelectronics      10,001+   
..       ...       ...                                    ...          ...   
254  Ireland   On-site                          permanent tsb  1,001-5,000   
255  Germany   On-site                               Sharpist      201-500   
256  Ireland   On-site                    TELUS International      10,001+   
257  Germany   On-site                       Rhenus Logistics      10,001+   
258   France   On-site                              Preligens       51-200   

                                            company_area  \
0   

In [51]:
df_fin = df_rel.explode('hard_skills')

In [52]:
df_fin.dtypes

index                    int64
html                    object
title                   object
place                   object
city                    object
country                 object
work_type               object
company_name            object
company_size            object
company_area            object
description             object
hard_skills             object
publish_date            object
candidats                Int64
date            datetime64[ns]
dtype: object

Оставим в датафрейме только необходимые столбцы

In [53]:
df_fin = df_fin[['index', 'title', 'city', 'country', 'work_type', 'company_name', 'company_size', 
                 'company_area', 'hard_skills', 'date', 'candidats']]

In [54]:
df_fin

index  \
0        0   
1        1   
2        2   
3        3   
4        4   
..     ...   
257    257   
258    258   
258    258   
258    258   
258    258   

                                                                     title  \
0                                                      Online Data Analyst   
1                                            Online Data Analyst - Belgium   
2                                                             Data Analyst   
3                                                             Data Analyst   
4                                   Alternant/ Alternante data analyst M/F   
..                                                                     ...   
257  Assistenz (w/m/d) des CEO West - Schwerpunkt Data Analyst / Contro...   
258                               Corporate Business Intelligence Engineer   
258                               Corporate Business Intelligence Engineer   
258                               Corporate Business Intelligence Engineer   
258                               Corporate Business Intelligence Engineer   

                           city  country work_type  \
0                         Skara   Sweden    Remote   
1                 West Flanders  Belgium    Remote   
2    Mecklenburg-West Pomerania  Germany    Remote   
3                       Hamburg  Germany    Remote   
4                       Rousset   France   On-site   
..                          ...      ...       ...   
257                  Düsseldorf  Germany   On-site   
258                       Paris   France   On-site   
258                       Paris   France   On-site   
258                       Paris   France   On-site   
258                       Paris   France   On-site   

                              company_name company_size  \
0    TELUS International AI Data Solutions      10,001+   
1                      TELUS International      10,001+   
2    TELUS International AI Data Solutions      10,001+   
3    TELUS International AI Data Solutions      10,001+   
4                       STMicroelectronics      10,001+   
..                                     ...          ...   
257                       Rhenus Logistics      10,001+   
258                              Preligens       51-200   
258                              Preligens       51-200   
258                              Preligens       51-200   
258                              Preligens       51-200   

                                            company_area hard_skills  \
0                          IT Services and IT Consulting         NaN   
1                          IT Services and IT Consulting         NaN   
2                          IT Services and IT Consulting         NaN   
3                          IT Services and IT Consulting         NaN   
4                            Semiconductor Manufacturing       excel   
..                                                   ...         ...   
257  Transportation, Logistics, Supply Chain and Storage        ssis   
258                      Defense and Space Manufacturing         api   
258                      Defense and Space Manufacturing         aws   
258                      Defense and Space Manufacturing       excel   
258                      Defense and Space Manufacturing         sql   

          date  candidats  
0   2022-09-01         12  
1   2022-09-01         11  
2   2022-09-07          2  
3   2022-09-07          4  
4   2022-09-02         25  
..         ...        ...  
257 2022-09-07       <NA>  
258 2022-09-01          2  
258 2022-09-01          2  
258 2022-09-01          2  
258 2022-09-01          2  

[802 rows x 11 columns]

In [55]:
df_fin.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 802 entries, 0 to 258
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   index         802 non-null    int64         
 1   title         802 non-null    object        
 2   city          771 non-null    object        
 3   country       802 non-null    object        
 4   work_type     588 non-null    object        
 5   company_name  802 non-null    object        
 6   company_size  802 non-null    object        
 7   company_area  729 non-null    object        
 8   hard_skills   728 non-null    object        
 9   date          802 non-null    datetime64[ns]
 10  candidats     701 non-null    Int64         
dtypes: Int64(1), datetime64[ns](1), int64(1), object(8)
memory usage: 76.0+ KB


Итак, после предобработки у нас получился датасет, состоящий из 802 строк и 10 столбцов с релевантными вакансиями, без дубликатов, с понятными датами публикации и корректными типами данных.

## Экспорт в csv

Подготовленный датасет сохраним в файл csv

In [56]:
df_fin.to_csv('linkedin_data.csv', index=False)